### PR TITLE
DEVPROD-3871 Don't consider hosts that are tearing down a group to be idle 

### PR DESCRIPTION
**Pull request by malikchaya2**
DEVPROD-3871

### Description
We use whether or not a host is running a task to determine if a host is idle. However, that is not sufficient because when assignNextTask sees that the task it's about to assign is in a different task group than the last task the ran, it will not set the running task and instead tell the agent to tear down the previous task group.

This can be seen with [this task](https://spruce.mongodb.com/task/evergreen_tests_oidc_test_oidc_azure_patch_841994da0546f3f0af36f9cb2150b33da32d7af5_65ef6e8661837d3882687ecd_24_03_11_20_50_23/logs?execution=0). Teardown group was set to sleep for ten minutes and the host was decommissioned and teardown group didn't run. 

### Testing
Added unit tests 
Tested on staging. The above task was run on staging with the changes. It slept for 5 minutes in the teardown group but it still ran the teardown group. 

note: this decommissioned the host _after_ running teardown, which will be fixed in [DEVPROD-5498](https://jira.mongodb.org/browse/DEVPROD-5498)

To test that hosts don't get stuck forever: on staging, temporarily set the default teardown timeout to 30 minutes and ran a task with a ten minute sleep. The host was successfully decommissioned before completing the teardown. 
[staging test linkt](https://spruce-staging.corp.mongodb.com/host/i-062e146e3cb537b0a)

### CODE REVIEW

**Comment by Kimchelly** [comment]
(This isn't a review, I just skimmed the changes really quick to understand the design. I'm going to be OOO soon so I'd rather someone else read this over if possible.)

First of all, nice job investigating this! I think the general idea of adding a check is sound. On top of the current changes, I think we probably want additional changes to ensure it's robust:

1. It looks from a glance like `IsTearingDown` functions similarly to `RunningTask` in that it's an flag indicating that the host being busy doing something useful. A lot of places in the codebase already rely on checking `RunningTask` as the marker of a host being idle or not. Now that there are two flags to indicate a host is busy doing task things (`RunningTask` and `IsTearingDown`), I imagine we'd have to update all the queries/checks involving `RunningTask` to also account for `IsTearingDown` as well.
2. I also have a related-ish thought that since `IsTearingDown` prevents a host from being shut down due to idleness, we have to take similar precautions to the ones we have in place for `RunningTask` to make sure that `IsTearingDown` doesn't potentially prevent a host from shutting down permanently in case it hits issues (e.g. somehow the agent gets stuck or OOMs while running teardown group, preventing it from ever unsetting the field). It seems like we could incorporate something like this into the [task timeout monitoring job](https://github.com/evergreen-ci/evergreen/blob/ced7e73f44f6958ed9ac8c2966269ac1321f4276/units/task_monitor_execution_timeout.go#L269-L273) to ensure that hosts don't get stuck running teardown group forever. Maybe we could add a hard (and fairly short) timeout limit on teardown group and say that if it's run for longer than this, the teardown group will fail?
3. Would it be worth writing an ADR for this? Teardown group is quite special because it's not considered part of a running task, so explaining what we're trying to work around and why it's needed specifically for teardown group may help us keep this knowledge for the future.

**Review by ablack12 — CHANGES_REQUESTED** [review]
Re Kim's comment, I think an ADR makes sense, and the concern around never shutting them down seems valid.

File: units/host_drawdown.go [hunk 1]
```diff
@@ -122,6 +122,11 @@ func (j *hostDrawdownJob) checkAndTerminateHost(ctx context.Context, h *host.Hos
 		return err
 	}
 
+	// don't drawdown hosts that are currently in the middle of tearing down a task group
+	if h.IsTearingDown {
```
[the remark this answers is not in the document]
Comment (malikchaya2) [reply units/host_drawdown.go]: We set isTearingDown before we return `nextTaskResponse.ShouldTeardownGroup = true` along with a nil task to the agent. If it reached the line that set isTearingDown to true, it's because it determined that a task group should be torn down and decided not to assign a new task. This is done at the same time as assignNextAvailableTask setting next task to nil, not on the next iteration.  

However, to avoid the concern of it potentially getting stuck in IsTearingDown forever I updated this check so it now takes into consideration the time it started.

File: model/host/db.go [hunk 2]
```diff
@@ -29,92 +29,97 @@ const (
...
+	VolumeTypeKey                          = bsonutil.MustHaveTag(Volume{}, "Type")
+	VolumeSizeKey                          = bsonutil.MustHaveTag(Volume{}, "Size")
+	VolumeExpirationKey                    = bsonutil.MustHaveTag(Volume{}, "Expiration")
+	VolumeNoExpirationKey                  = bsonutil.MustHaveTag(Volume{}, "NoExpiration")
+	VolumeHostKey                          = bsonutil.MustHaveTag(Volume{}, "Host")
+	VolumeMigratingKey                     = bsonutil.MustHaveTag(Volume{}, "Migrating")
+	VolumeAttachmentIDKey                  = bsonutil.MustHaveTag(VolumeAttachment{}, "VolumeID")
+	VolumeDeviceNameKey                    = bsonutil.MustHaveTag(VolumeAttachment{}, "DeviceName")
+	DockerOptionsStdinDataKey              = bsonutil.MustHaveTag(DockerOptions{}, "StdinData")
+	SleepScheduleNextStopTimeKey           = bsonutil.MustHaveTag(SleepScheduleInfo{}, "NextStopTime")
```
Comment (ablack12) [inline model/host/db.go:118]: Was this a weird merge? This seems unrelated to your changes.

File: model/host/host.go [hunk 3]
```diff
@@ -982,39 +988,40 @@ func (h *Host) SetAgentStartTime(ctx context.Context) error {
 	return nil
 }
 
-// SetIsTearingDown sets the IsTearingDown flag for the host
-func (h *Host) SetIsTearingDown(ctx context.Context) error {
+// SetTaskGroupTeardownStartTime sets the TaskGroupTeardownStartTime to the current time for the host
+func (h *Host) SetTaskGroupTeardownStartTime(ctx context.Context) error {
```
Comment (ablack12) [inline model/host/host.go:992]: Can we add unit tests for these?

File: model/host/host.go [hunk 4]
```diff
@@ -2037,6 +2044,11 @@ func (h *Host) GetElapsedCommunicationTime() time.Duration {
 	return time.Since(h.CreationTime)
 }
 
+// GetElapsedTeardownTime calculates the elapsed time since the teardown of the task group started.
+func (h *Host) GetElapsedTeardownTime() time.Duration {
```
Comment (ablack12) [inline model/host/host.go:2048]: since the only place we use this we check the value against a global variable, could we just rename this be like "TeardownTimeExceeded" or something, and move that check in here? (i.e. `time.Since(h.TaskGroupTeardownStartTime) > evergreen.MaxTeardownGroupThreshold`)

**Review by ablack12 — COMMENTED** [review]
I like this change!

File: globals.go [hunk 5]
```diff
@@ -363,6 +363,10 @@ const (
 	// a heartbeat
 	HeartbeatTimeoutThreshold = 7 * time.Minute
 
+	// MaxTeardownGroupThreshold specifies the duration after which the host should no longer continue
+	// to tear down a task group. This is set one minute longer than the agent's maxTeardownGroupTimeout.
+	MaxTeardownGroupThreshold = 4 * time.Minute
```
[the remark this answers is not in the document]
Comment (malikchaya2) [reply globals.go:368]: This is longer because this is for the case where something happens and the host gets stuck. In an ideal environment the timeout will be what tears it down. This should only kick in if something went really wrong, so setting it to 4 minutes gives it a buffer.

File: model/host/db.go (hunk 2 above)
Comment (malikchaya2) [reply model/host/db.go:118]: weird, I don't see this in the files tab, were you looking at the merge commit directly?

Comment (malikchaya2) [reply model/host/db.go:118]: It looks like it's part of [this merge ](https://github.com/evergreen-ci/evergreen/commit/a12320581a042adfa2ba3c0793ad2b14e4839110)

File: docs/decisions/2024-03-26_teardown_group_idleness.md [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+Another approach to address this issue is to record the start time of the teardown group on the host and utilize it to determine the host's state.
+
+## Decision Outcome
+
+The solution was to add a `TaskGroupTeardownStartTime` field to hosts that sets the time when `assignNextTask` sends back a boolean to the agent indicating that it should tear down the task group. A maximum teardown group threshold of 3 minutes was implemented. The user-configurable teardown group timeout will still be respected, but only if it is less than the maximum teardown group threshold. The choice of 3 minutes aligns with the original intent of teardown tasks being used for short operations while allowing some flexibility for special case scenarios.
```
[the remark this answers is not in the document]
Comment (malikchaya2) [reply docs/decisions/2024-03-26_teardown_group_idleness.md:23]: This will not impact existing flows based on all-configs because the only project that has it set longer than 2 minutes (at 30 minutes) is the project that reported this bug, presumably to try to fix it. 
That being said, this is not comprehensive enough because you can add task groups in generate.tasks so I plan to send out some comms after this is merged (and add a comment to the ticket for the affected project).

File: model/host/db.go (hunk 2 above)
Comment (malikchaya2) [reply model/host/db.go]: I realized what was happening. When you filter commits to "since your last review", if there is a merge in there, that is shown as well.  
image

Comment (ablack12) [reply model/host/db.go]: Oh cool -- i very rarely use this setting so I didn't consider that. Thanks!

File: validator/project_validator.go [hunk 7]
```diff
@@ -1767,6 +1767,13 @@ func checkTaskGroups(p *model.Project) ValidationErrors {
 		}
 	}
 	for _, tg := range taskGroups {
+		// validate that teardown group timeout is not over MaxTeardownGroupTimeout
+		if tg.TeardownGroupTimeoutSecs > int(globals.MaxTeardownGroupTimeout.Seconds()) {
+			errs = append(errs, ValidationError{
+				Message: fmt.Sprintf("task group %s has a teardown task timeout of %d seconds, which exceeds the maximum of %d seconds", tg.Name, tg.TeardownGroupTimeoutSecs, int(globals.MaxTeardownGroupTimeout.Seconds())),
```
Comment (ablack12) [inline validator/project_validator.go:1773]: nit: '%s'

File: model/host/host_test.go [hunk 8]
```diff
@@ -762,6 +762,29 @@ func TestHostSetAgentStartTime(t *testing.T) {
...
+	}()
+
+	h := &Host{
+		Id: "id",
+	}
+	require.NoError(t, h.Insert(ctx))
+
+	now := time.Now()
+	require.NoError(t, h.SetTaskGroupTeardownStartTime(ctx))
+	assert.True(t, now.Sub(h.TaskGroupTeardownStartTime) < time.Second)
```
Comment (ablack12) [inline model/host/host_test.go:781]: Can we assert this is false before making the call, just to verify?

**Review by ablack12 — APPROVED** [review]
🚀
